source/extensions/filters/network/network_level_sni_reader/network_level_sni_reader.cc

```diff
@@ -0,0 +1,89 @@
+#include "extensions/filters/network/network_level_sni_reader/network_level_sni_reader.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+#include "envoy/network/connection.h"
+#include "envoy/stats/scope.h"
+
+#include "common/common/assert.h"
+
+#include "openssl/bytestring.h"
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace NetworkLevelSniReader {
+
+Config::Config(Stats::Scope& scope, uint32_t max_client_hello_size)
+    : stats_{TLS_STATS(POOL_COUNTER_PREFIX(scope, "network_levelsni_reader."))},
+      ssl_ctx_(SSL_CTX_new(TLS_with_buffers_method())),
+      max_client_hello_size_(max_client_hello_size) {
+
+  if (max_client_hello_size_ > TLS_MAX_CLIENT_HELLO) {
+    throw EnvoyException(fmt::format("max_client_hello_size of {} is greater than maximum of {}.",
+                                     max_client_hello_size_, size_t(TLS_MAX_CLIENT_HELLO)));
+  }
+
+  SSL_CTX_set_options(ssl_ctx_.get(), SSL_OP_NO_TICKET);
+  SSL_CTX_set_session_cache_mode(ssl_ctx_.get(), SSL_SESS_CACHE_OFF);
+  SSL_CTX_set_tlsext_servername_callback(
```

Comment:
If we do go with this PR, can you factor out as much common stuff with the TLS inspector as possible? It might only be a small amount of code, but I would rather we have a single place to fix/maintain/audit. Thanks.

Reply:
Sure, we will refactor the code to create as less code duplication as possible, and will follow all the code review remarks and the requirements of Envoy.